doc/md: guide on deploying versioned migrations (#1247)
ariga/atlas: doc/md/guides/deploying/image.md

---
id: image
slug: /guides/deploying/image
title: Creating container images for migrations
---

To integrate schema migrations into pipelines that deploy to container management
systems (such as Kubernetes, AWS ECS, Google Cloud Run, etc.) it is recommended
to create a dedicated container image per version that contains the
migration tool (such as Atlas) and the relevant migration files.

In this guide we will demonstrate how to build a dedicated Docker image that

> includes Atlas and the relevant migrations files. We will demonstrate

Reviewer: ```suggestion
includes Atlas and its migration files. We will demonstrate
```

how to build this image as a [GitHub Actions Workflow](https://docs.github.com/en/actions/using-workflows),
but the same result can be achieved in any CI system. 

## Defining the Dockerfile

Suppose our project structure looks something like: 

```text
.
├── main.go
└── migrations
    ├── 20221031125934_init.sql
    ├── 20221031125940_add_users_table.sql
    ├── 20221031125948_add_products_table.sql
    └── atlas.sum
```

Our goal is to build an image that contains:
1. The `migrations` directory
2. The Atlas binary

> To do this we can build our container image shall with the official [Atlas Docker image](https://hub.docker.com/r/arigaio/atlas)

Reviewer: ```suggestion
To do this we can build our container image shell with the official [Atlas Docker image](https://hub.docker.com/r/arigaio/atlas)
```
You meant 'shell', right? :)

Author: Neither 🤗

as [the base layer](https://docs.docker.com/engine/reference/builder/#from).  

> To do this our Dockerfile should be placed in the directory _containing_ the `migrations`

Reviewer: ```suggestion
To do this, our Dockerfile should be placed in the directory _containing_ the `migrations`
```

> directory and look something like this:

Reviewer: ```suggestion
directory and will look something like this:
```

 => => transferring dockerfile: 36B                                                           0.0s
 => [internal] load .dockerignore                                                             0.0s
 => => transferring context: 2B                                                               0.0s
 => [internal] load metadata for docker.io/arigaio/atlas:latest                               0.0s
 => [internal] load build context                                                             0.0s
 => => transferring context: 252B                                                             0.0s
 => [1/2] FROM docker.io/arigaio/atlas:latest                                                 0.0s
 => CACHED [2/2] COPY migrations /migrations                                                  0.0s
 => exporting to image                                                                        0.0s
 => => exporting layers                                                                       0.0s
 => => writing image sha256:c928104de31fc4c99d114d40ea849ade917beae3df7ffe9326113b289939878e  0.0s
 => => naming to docker.io/library/my-image                                                   0.0s
```

> To verify Atlas can find your migrations directory that it's [integrity](/concepts/migration-directory-integrity)

Reviewer: ```suggestion
To verify Atlas can find your migrations directory and that its [integrity](/concepts/migration-directory-integrity)
```

is intact run:

```text
docker run --rm migrations my-image migrate validate 
```

If no issues are found, no errors will be printed out.

## Defining the GitHub Actions Workflow

Next, we define a GitHub Actions workflow that will build our container

> image and push it to GitHub container repo (ghcr.io) on every push

Reviewer: ```suggestion
image and push it to the GitHub container repo (ghcr.io) on every push
```
